internal/impl/redpanda/processor_data_transforms.go

```diff
@@ -0,0 +1,414 @@
+	if err != nil {
+		_ = modRunner.Close(ctx)
+		return nil, err
+	}
+	p.modulePool.Put(modRunner)
+	return []service.MessageBatch{res}, nil
+}
+
+func (p *dataTransformEnginePool) Close(ctx context.Context) error {
+	for {
+		mr := p.modulePool.Get()
+		if mr == nil {
+			return nil
+		}
+		if err := mr.(*dataTransformEngine).Close(ctx); err != nil {
+			return err
+		}
+	}
+}
+
+//------------------------------------------------------------------------------
+
+type dataTransformEngine struct {
+	log *service.Logger
+	cfg *dataTransformConfig
+
+	runtime wazero.Runtime
+	mod     api.Module
+
+	inputBatch  []transformMessage
+	outputBatch service.MessageBatch
+	targetIndex int
+
+	procErr   error
+	hostChan  chan any
+	guestChan chan any
+}
+
+func (r *dataTransformEngine) newTransformMessage(message *service.Message) (tmsg transformMessage, err error) {
+	tmsg.value, err = message.AsBytes()
+	if err != nil {
+		return
+	}
+	if r.cfg.inputKey != nil {
+		if tmsg.key, err = r.cfg.inputKey.TryBytes(message); err != nil {
+			return
+		}
+	}
+	if r.cfg.timestamp != nil {
+		var tsStr string
+		if tsStr, err = r.cfg.timestamp.TryString(message); err != nil {
+			err = fmt.Errorf("timestamp interpolation error: %w", err)
+			return
+		}
+		if tmsg.timestamp, err = strconv.ParseInt(tsStr, 10, 64); err != nil {
+			err = fmt.Errorf("failed to parse timestamp: %w", err)
+			return
+		}
+	} else {
+		tmsg.timestamp = time.Now().UnixMilli()
+	}
+	err = r.cfg.inputMetadata.Walk(message, func(key string, value string) error {
+		tmsg.headers = append(tmsg.headers, transformHeader{key, []byte(value)})
+		return nil
+	})
+	return
+}
+
+func (r *dataTransformEngine) convertTransformMessage(message transformMessage) (*service.Message, error) {
+	msg := service.NewMessage(message.value)
+	if r.cfg.outputMetadata != nil {
+		for _, hdr := range message.headers {
+			if r.cfg.outputMetadata.Match(hdr.key) {
+				msg.MetaSetMut(hdr.key, hdr.value)
+			}
+		}
+	}
+	if r.cfg.outputKeyField != nil {
+		msg.MetaSetMut(*r.cfg.outputKeyField, message.key)
+	}
+	if message.outputTopic != nil {
+		msg.MetaSetMut("data_transform_output_topic", *message.outputTopic)
+	}
+	return msg, nil
+}
+
+func (r *dataTransformEngine) reset() {
+	r.inputBatch = nil
+	r.targetIndex = 0
+	r.outputBatch = nil
+}
+
+func (r *dataTransformEngine) Run(ctx context.Context, batch service.MessageBatch) (service.MessageBatch, error) {
+	if r.procErr != nil {
+		return nil, r.procErr
+	}
+	defer r.reset()
+	r.inputBatch = make([]transformMessage, len(batch))
+	r.targetIndex = 0
+	for i, msg := range batch {
+		tm, err := r.newTransformMessage(msg)
+		if err != nil {
+			return nil, err
+		}
+		r.inputBatch[i] = tm
+	}
+	// Notify the guest that it has data to process
+	r.guestChan <- nil
+	// Wait for the guest to process everything
+	select {
+	case <-r.hostChan:
+	case <-time.After(r.cfg.timeout):
+		_ = r.mod.Close(ctx)
+		drainChannel(r.hostChan)
+	}
+	return r.outputBatch, r.procErr
```

Comment:
I noticed the other module called SetError on each individual message and never returned an error here - is that a better practice? I'm assuming this will stop the pipeline while the service.Message.SetError way does not?

Reply:
I think it's covered in the docs: https://pkg.go.dev/github.com/redpanda-data/benthos/v4@v4.31.0/public/service#BatchProcessor

> // Process a batch of messages into one or more resulting batches, or return
> // an error if the entire batch could not be processed. If zero messages are
> // returned and the error is nil then all messages are filtered.
> //
> // The provided MessageBatch should NOT be modified, in order to return a
> // mutated batch a copy of the slice should be created instead.
> //
> // When an error is returned all of the input messages will continue down
> // the pipeline but will be marked with the error with *message.SetError,
> // and metrics and logs will be emitted.
> //
> // In order to add errors to individual messages of the batch for downstream
> // handling use *message.SetError(err) and return it in the resulting batch
> // with a nil error.
> //
> // The Message types returned MUST be derived from the provided messages,
> // and CANNOT be custom instantiations of Message. In order to copy the
> // provided messages use the Copy method.

Sometimes, I find it useful to set a breakpoint here and follow the call stack. There's some logic here: https://github.com/redpanda-data/benthos/blob/d940a3e6243d9c6e615fa57f99bc90f55d360c22/public/service/processor.go#L112-L115 and then here https://github.com/redpanda-data/benthos/blob/c209d989cf506fcb25c5ac0b0aa1ca8e90f54c09/internal/component/processor/auto_observed.go#L223 which shows what's going on. If you return an error, the returned batch is discarded in the `airGapBatchProcessor` and then the `AutoObservedBatched` processor will take the original input batch, mark all the messages in it with the received error and then return this batch. Not sure what you mean by stopping the pipeline, but you can filter out messages if you don't include them in the returned batch (and also return a nil error).

Reply:
Ah thank you for the detailed explaination. I believe applying an error to the entire batch is the correct approach, the ABI works and checkpoints per batch.